consensus/ethash/consensus.go

```diff
@@ -0,0 +1,495 @@
+					continue
+				}
+				// We need to look up the first parent
+				var parent *types.Header
+				if index == 0 {
+					parent = chain.GetHeader(headers[0].ParentHash, headers[0].Number.Uint64()-1)
+				} else if headers[index-1].Hash() == headers[index].ParentHash {
+					parent = headers[index-1]
+				}
+				// Ensure the validation is useful and execute it
+				var failure error
+				switch {
+				case chain.GetHeader(headers[index].Hash(), headers[index].Number.Uint64()-1) != nil:
+					outputs <- result{index: index, err: nil}
+				case parent == nil:
+					failure = ErrParentUnknown
+					outputs <- result{index: index, err: failure}
+				default:
+					failure = ethash.verifyHeader(chain, headers[index], parent, false, seals[index])
+					outputs <- result{index: index, err: failure}
+				}
+				// If a validation failure occurred, mark subsequent blocks invalid
+				if failure != nil {
+					number := headers[index].Number.Uint64()
+					if prev := atomic.LoadUint64(&badblock); prev == 0 || prev > number {
+						// This two step atomic op isn't thread-safe in that `badblock` might end
+						// up slightly higher than the block number of the first failure (if many
+						// workers try to write at the same time), but it's fine as we're mostly
+						// interested to avoid large useless work, we don't care about 1-2 extra
+						// runs. Doing "full thread safety" would involve mutexes, which would be
+						// a noticeable sync overhead on the fast spinning worker routines.
+						atomic.StoreUint64(&badblock, number)
+					}
+				}
+			}
+		}()
+	}
+	// Feed item indices to the workers until done, sorting and feeding the results to the caller
+	dones := make([]bool, len(headers))
+	errors := make([]error, len(headers))
+
+	abort := make(chan struct{})
+	returns := make(chan error, len(headers))
+
+	go func() {
+		defer close(inputs)
+
+		input, output := 0, 0
+		for i := 0; i < len(headers)*2; i++ {
+			var res result
+
+			// If there are tasks left, push to workers
+			if input < len(headers) {
+				select {
+				case inputs <- input:
+					input++
+					continue
+				case <-abort:
+					return
+				case res = <-outputs:
+				}
+			} else {
+				// Otherwise keep waiting for results
+				select {
+				case <-abort:
+					return
+				case res = <-outputs:
+				}
+			}
+			// A result arrived, save and propagate if next
+			dones[res.index], errors[res.index] = true, res.err
+			for output < len(headers) && dones[output] {
+				returns <- errors[output]
+				output++
+			}
+		}
+	}()
+	return abort, returns
+}
+
+// VerifyUncles verifies that the given block's uncles conform to the consensus
+// rules of the stock Ethereum ethash engine.
+func (ethash *Ethash) VerifyUncles(chain consensus.ChainReader, block *types.Block) error {
+	// If we're running a full engine faking, accept any input as valid
+	if ethash.fakeFull {
+		return nil
+	}
+	// Verify that there are at most 2 uncles included in this block
+	if len(block.Uncles()) > 2 {
```

Comment:
I was also thinking about this. Though perhaps I'd put it into a params.go inside the `ethash` package, since uncles are only meaningful for ethash, and not for anything else (clique currently, but probably casper too).